### PR TITLE
Detect changes/drift through a filtered diff of the apply

### PR DESCRIPTION
**Pull request by JaneLiuL**
This PR including below contents:
1. run kubectl diff first and parse the output. .
2. update the object events.
3. test for the output

Detail could be check on: https://github.com/fluxcd/kustomize-controller/issues/352

### CODE REVIEW

**Comment by stefanprodan** [comment]
We need a test that proves that we ignore objects that are changed only due to the checksum annotation. We also need to test diff failures, for example a namespace and configmap will error out on create, but will succeed on update.

**Comment by JaneLiuL** [comment]
Get ~ I will add it now~

**Comment by hiddeco** [comment]
I think that as a future optimization it would also be an idea to create some type of wrapper around the working directory and/or commands we run so that we can omit the `cd ... &&` (or make them simple methods that run specific commands, or something something).

File: controllers/utils.go [hunk 1]
```diff
@@ -87,3 +154,15 @@ func ObjectKey(object metav1.Object) client.ObjectKey {
 		Name:      object.GetName(),
 	}
 }
+
+func containsChangeInSlice(tmpslice []string) bool {
+	for _, s := range tmpslice {
+		if strings.HasPrefix(s, "+    kustomize.toolkit.fluxcd.io/checksum:") || strings.HasPrefix(s, "-    kustomize.toolkit.fluxcd.io/checksum:") || strings.HasPrefix(s, "--- /tmp") || strings.HasPrefix(s, "+++ /tmp") {
```
Comment (stefanprodan) [inline controllers/utils.go:160]: Please declare a var above like `checksumAnnotation := fmt.Sprintf("%s/checksum", kustomizev1.GroupVersion.Group)`

**Comment by stefanprodan** [comment]
Some things to note about this PR:
* if the diff returns no changeset, then the controller skips the apply
* the diff excludes changes to the checksum annotation trying to reduce the event spam but it fails to do so for Kubernetes Deployments, Statefulsets and Daemonsets due to the fact that an annotation change bumps the object generation 
* the diff errors out with exit code 2 if newly created objects are part of a new namespace
* we ignore the diff crash and fallback to the apply output to compose the changeset
* the apply output is very different to the diff one, users will receive inconsistent events, when diff fails the changeset format is `<object-name> <action>`, if diff works then the format is `<api-version>.<kind>.<namespace>.<name> <action>`
* we have no idea about the performance impact of this change, in theory this would double the number of API calls made to the API server

File: controllers/kustomization_controller.go [hunk 2]
```diff
@@ -645,28 +650,41 @@ func (r *KustomizationReconciler) apply(ctx context.Context, kustomization kusto
...
-
-		applyErr := parseApplyError(output)
-		if applyErr == "" {
-			applyErr = "no error output found, this may happen because of a timeout"
-		}
-		return "", fmt.Errorf("apply failed: %s", applyErr)
-	}
-
-	resources := parseApplyOutput(output)
 	log.Info(
```
Comment (stefanprodan) [inline controllers/kustomization_controller.go:688]: We shouldn't log that we applied something on the cluster if we didn't, please move the log below inside `else if differr.ExitCode() == 1 `

File: controllers/kustomization_controller.go [hunk 3]
```diff
@@ -645,28 +650,41 @@ func (r *KustomizationReconciler) apply(ctx context.Context, kustomization kusto
...
+	// For the "diff" command Exit status:
+	//		 0
+	//		No differences were found.
+	//		 1
+	//		Differences were found.
+	//		 >1
+	//		Kubectl or diff failed with an error.
+	if differr, ok := err.(*exec.ExitError); ok {
+		if differr.ExitCode() > 1 {
+			log.Info("Kubectl or diff failed with an error, will execute apply soon")
```
Comment (stefanprodan) [inline controllers/kustomization_controller.go:671]: Make this a debug log please.

File: controllers/kustomization_controller.go [hunk 4]
```diff
@@ -645,28 +650,41 @@ func (r *KustomizationReconciler) apply(ctx context.Context, kustomization kusto
...
+			output, applyErr := execApply(ctx, cmd)
+			if applyErr != nil {
+				return "", applyErr
+			}
+			resources = parseApplyOutput(output)
+		} else if differr.ExitCode() == 1 {
+			resources = parseDiffOutput(diffoutput)
+			//	Since we found difference in "diff" command, so we need to execute apply
+			log.Info("Differences found, will execute apply soon")
```
Comment (stefanprodan) [inline controllers/kustomization_controller.go:680]: Make this a debug log please.

**Comment by stefanprodan** [comment]
One unfortunate outcome of skipping the apply when the diff returns no changes is that enabling pruning will not be possible, as the the annotations will never be added.

File: controllers/kustomization_controller.go (hunk 4 above)
Comment (JaneLiuL) [reply controllers/kustomization_controller.go]: Fix done

File: controllers/kustomization_controller.go (hunk 3 above)
Comment (JaneLiuL) [reply controllers/kustomization_controller.go]: Fix done~

File: controllers/kustomization_controller.go (hunk 2 above)
Comment (JaneLiuL) [reply controllers/kustomization_controller.go:689]: Fix done

File: controllers/utils.go (hunk 1 above)
Comment (JaneLiuL) [reply controllers/utils.go]: Fix done

**Comment by JaneLiuL** [comment]
Any update for this PR? Should I still keep this PR or not? @stefanprodan

**Comment by hiddeco** [comment]
:wave: 

We have not forgotten about this PR, but are currently in the process of a bigger refactor which will drastically change the structure of all controller's reconcilers (see https://github.com/fluxcd/flux2/issues/1601 for details), and improve the test coverage of the individual operations a controller performs.

Given one of the goals of this PRs is to hypothetically improve performance, we think it is better for the refactor to land first, so that we are properly able to benchmark your proposal against the current way the controller does things.

Well aware this may be a bummer to you, and I am really sorry about that. :sunflower:

**Comment by JaneLiuL** [comment]
Close this PR since already have https://github.com/stefanprodan/kustomizer/pull/17 to cove

**Pull request closed without merging**